wallet/walletrpc.c

```diff
@@ -316,10 +316,28 @@ static void json_newaddr(struct command *cmd,
 	struct json_result *response = new_json_result(cmd);
 	struct ext_key ext;
 	struct sha256 h;
-	struct ripemd160 p2sh;
+	struct ripemd160 h160;
 	struct pubkey pubkey;
+	jsmntok_t *addrtype;
 	u8 *redeemscript;
+	bool is_p2wpkh, ok;
 	s64 keyidx;
+	char *out;
+
+	if (!json_get_params(buffer, params, "?addresstype", &addrtype, NULL)) {
+		command_fail(cmd, "Invalid arguments");
+		return;
+	}
+
+	if (!addrtype || json_tok_streq(buffer, addrtype, "p2sh-segwit"))
+		is_p2wpkh = false;
+	else if (json_tok_streq(buffer, addrtype, "bech32"))
+		is_p2wpkh = true;
+	else {
+		command_fail(cmd,
+			     "Invalid address type param (expected segwit/p2wpkh or p2sh)");
```

Comment:
Error message is misleading --- it seems from the error message that you accept "segwit/p2wpkh" or "p2sh", but really you accept instead "p2sh-segwit" or "bech32").

Reply:
Good catch!